### PR TITLE
Native promises

### PR DESCRIPTION
**Pull request by THoisington**
implements native promises (both explicitly and with promisify). Refactors to remove When.js and lodash. Updates node version to v8.0.0. Addresses issue #13

### CODE REVIEW

File: lib/request.js [hunk 1]
```diff
@@ -14,37 +12,44 @@ const version = require("../package.json").version;
...
 
     if (baseConfig.auth) {
-        config.auth = _.assign({}, baseConfig.auth);
-        config.auth.sendImmediately = "sendImmediately" in baseConfig.auth ? baseConfig.auth.sendImmediately : false;
+        config.auth = {
+            sendImmediately:
+                "sendImmediately" in baseConfig.auth ?
+                    baseConfig.auth.sendImmediately :
+                    false,
+            ...baseConfig.auth
```
Comment (meaku) [inline lib/request.js]: We should get the same result as default with spread: 

```
config.auth = {
   sendImmediately: false, //default
   ...baseConfig.auth //overwrites
}
```

That's basically what the expression `"sendImmediately" in baseConfig.auth ? baseConfig.auth.sendImmediately : false,` does, right?

File: lib/request.js [hunk 2]
```diff
@@ -14,37 +12,44 @@ const version = require("../package.json").version;
...
+    return new Promise((resolve, reject) => {
+        doRequest(config, (res, body) => {
             let err;
 
             if (res.statusCode >= 300) {
                 err = new Error("Response with status code: " + res.statusCode);
                 err.statusCode = res.statusCode;
-                throw err;
+                reject(err);
+            } else {
```
Comment (meaku) [inline lib/request.js]: I personally prefer something like the following to reduce braces, but it's up to you. 

```javascript
if(res.statusCode >= 300){
    reject(err);
    return; 
} 

resolve(body); 
```

File: lib/request.js [hunk 3]
```diff
@@ -14,37 +12,44 @@ const version = require("../package.json").version;
...
-    config.body = xml;
-    config.method = method;
-    config.headers["Content-length"] = xml.length;
-    config.headers.Depth = depth;
     config.timeout = baseConfig.timeout ? baseConfig.timeout : 10000;
 
-    return nodefn.call(doRequest, config)
-        .spread((res, body) => {
+    return new Promise((resolve, reject) => {
+        doRequest(config, (res, body) => {
```
Comment (meaku) [inline lib/request.js]: I was wondering if there is an error argument as well? What happens if request messes up really badly and won't even return a valid `res` at all?

File: .travis.yml [hunk 4]
```diff
@@ -1,6 +1,8 @@
 language: node_js
 node_js:
   - "6"
```
Comment (meaku) [inline .travis.yml]: We have to remove version 6 here.

Comment (meaku) [reply .travis.yml]: And maybe add 10 :)

**Comment by THoisington** [comment]
- [ ] Test with WWW dev using npm link

File: lib/request.js [hunk 5]
```diff
@@ -14,37 +12,48 @@ const version = require("../package.json").version;
...
 function request(baseConfig, method, depth, xml) {
-    const config = _.assign({}, baseConfig);
-
-    config.headers = _.assign({}, baseConfig.headers);
+    const config = {
+        headers: baseConfig.headers || {},
+        body: xml,
+        method,
+        'headers["Content-length"]': xml.length,
+        'headers["Depth"]': depth,
```
Comment (antosan) [inline lib/request.js]: This will not add the headers correctly to the config. We should add them directly to the headers object, like this:

**Comment by THoisington** [comment]
@antosan Can you take another look?

**Pull request merged**